File: mobile/lib/models/course.dart
class Course {
  final int id;
  final String title;
  final String description;
  final String category;
  final String level;
  final String? coverImageUrl;
  final int createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;
  final double? averageRating;
  final int reviewCount;

  Course({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.level,
    required this.coverImageUrl,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
    this.averageRating,
    this.reviewCount = 0,
  });

  factory Course.fromJson(Map<String, dynamic> json) => Course(
    id: json['id'] as int,
    title: json['title'] as String,
    description: json['description'] as String,
    category: json['category'] as String,
    level: json['level'] as String,
    coverImageUrl: json['cover_image_url'] as String?,
    createdBy: json['created_by'] as int,
    createdAt: DateTime.parse(json['created_at'] as String),
    updatedAt: DateTime.parse(json['updated_at'] as String),
    averageRating: (json['average_rating'] as num?)?.toDouble(),
    reviewCount: json['review_count'] as int? ?? 0,
  );
}

class CourseDetail extends Course {
  final int lessonCount;

  CourseDetail({
    required super.id,
    required super.title,
    required super.description,
    required super.category,
    required super.level,
    required super.coverImageUrl,
    required super.createdBy,
    required super.createdAt,
    required super.updatedAt,
    super.averageRating,
    super.reviewCount = 0,
    required this.lessonCount,
  });

  factory CourseDetail.fromJson(Map<String, dynamic> json) => CourseDetail(
    id: json['id'] as int,
    title: json['title'] as String,
    description: json['description'] as String,
    category: json['category'] as String,
    level: json['level'] as String,
    coverImageUrl: json['cover_image_url'] as String?,
    createdBy: json['created_by'] as int,
    createdAt: DateTime.parse(json['created_at'] as String),
    updatedAt: DateTime.parse(json['updated_at'] as String),
    averageRating: (json['average_rating'] as num?)?.toDouble(),
    reviewCount: json['review_count'] as int? ?? 0,
    lessonCount: json['lesson_count'] as int,
  );
}

File: backend/app/schemas/course.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, field_validator


class CourseCreate(BaseModel):
    title: str
    description: str = ""
    category: str = ""
    level: str = "beginner"
    cover_image_url: str | None = None

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Title is required")
        return v.strip()


class CourseUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    category: str | None = None
    level: str | None = None
    cover_image_url: str | None = None


class CourseOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str
    category: str
    level: str
    cover_image_url: str | None
    created_by: int
    created_at: datetime
    updated_at: datetime
    average_rating: float | None = None
    review_count: int = 0


class CourseDetailOut(CourseOut):
    lesson_count: int
